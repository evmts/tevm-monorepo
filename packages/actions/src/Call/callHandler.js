import { createAddress } from '@tevm/address'
import { numberToBytes } from 'viem'
import { createScript } from '../Contract/createScript.js'
import { getL1FeeInformationOpStack } from '../internal/getL1FeeInformationOpStack.js'
import { getPendingClient } from '../internal/getPendingClient.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { callHandlerOpts } from './callHandlerOpts.js'
import { callHandlerResult } from './callHandlerResult.js'
import { cloneVmWithBlockTag } from './cloneVmWithBlock.js'
import { executeCall } from './executeCall.js'
import { handlePendingTransactionsWarning } from './handlePendingTransactionsWarning.js'
import { handleTransactionCreation } from './handleTransactionCreation.js'
import { validateCallParams } from './validateCallParams.js'

/**
 * Creates a tree-shakable instance of [`client.tevmCall`](https://tevm.sh/reference/tevm/decorators/type-aliases/tevmactionsapi/#call) action.
 * This function is designed for use with TevmNode and the internal instance of TEVM,
 * and it is distinct from the viem API `tevmCall`.
 *
 * Note: This is the internal logic used by higher-level APIs such as `tevmCall`.
 *
 * @param {import('@tevm/node').TevmNode} client - The TEVM base client instance.
 * @param {object} [options] - Optional parameters.
 * @param {boolean} [options.throwOnFail=true] - Whether to throw an error on failure.
 * @returns {import('./CallHandlerType.js').CallHandler} The call handler function.
 * @throws {import('./TevmCallError.js').TevmCallError} If `throwOnFail` is true, returns `TevmCallError` as value.
 *
 * @example
 * ```typescript
 * import { createTevmNode } from 'tevm/node'
 * import { callHandler } from 'tevm/actions'
 *
 * const client = createTevmNode()
 *
 * const call = callHandler(client)
 *
 * const res = await call({
 *   createTransaction: true,
 *   to: `0x${'69'.repeat(20)}`,
 *   value: 420n,
 *   skipBalance: true,
 * })
 * ```
 */
export const callHandler =
	(client, { throwOnFail: defaultThrowOnFail = true } = {}) =>
	async ({ code, deployedBytecode, ...params }) => {
		/**
		 * ***************
		 * 0 VALIDATE PARAMS
		 * ***************
		 */
		client.logger.debug(params, 'callHandler: Executing call with params')
		const validationErrors = validateCallParams(params)
		if (validationErrors.length > 0) {
			client.logger.debug(validationErrors, 'Params do not pass validation')
			return maybeThrowOnFail(params.throwOnFail ?? defaultThrowOnFail, {
				errors: validationErrors,
				executionGasUsed: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				rawData: '0x',
			})
		}
		const _params = {
			...params,
		}

		if (_params.blockTag === 'pending') {
			const minePending = await getPendingClient(client)
			if (minePending.errors) {
				client.logger.error(minePending.errors)
				return maybeThrowOnFail(_params.throwOnFail ?? defaultThrowOnFail, {
					errors: minePending.errors,
					executionGasUsed: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					rawData: '0x',
				})
			}
			// if we are creating a transaction we want to use the real txpool so the tx gets properly added
			if (_params.createTransaction) {
				const pendingClientAny = /** @type {any}*/ (minePending.pendingClient)
				pendingClientAny.getTxPool = client.getTxPool
			}
			return callHandler(minePending.pendingClient, { throwOnFail: defaultThrowOnFail })({
				...(code !== undefined ? { code } : {}),
				...(deployedBytecode !== undefined ? { deployedBytecode } : {}),
				..._params,
				blockTag: 'latest',
			})
		}

		const callHandlerRes = await callHandlerOpts(client, _params)
		if (callHandlerRes.errors) {
			return maybeThrowOnFail(_params.throwOnFail ?? defaultThrowOnFail, {
				errors: callHandlerRes.errors,
				executionGasUsed: 0n,
				rawData: /** @type {`0x${string}`}*/ ('0x'),
			})
		}

		const evmInput = callHandlerRes.data

		const block = /** @type {import('@tevm/block').Block}*/ (evmInput.block)

		await handlePendingTransactionsWarning(client, params, code, deployedBytecode)

		/**
		 * ************
		 * 1 CLONE THE VM WITH BLOCK TAG
		 * ************
		 */
		const vm = await cloneVmWithBlockTag(client, block)
		if (vm instanceof Error) {
			return maybeThrowOnFail(_params.throwOnFail ?? defaultThrowOnFail, {
				errors: [vm],
				executionGasUsed: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				rawData: '0x',
			})
		}

		const scriptResult =
			code || deployedBytecode
				? await createScript({ ...client, getVm: () => vm.ready().then(() => vm) }, code, deployedBytecode, _params.to)
				: { address: _params.to, errors: undefined }
		if (scriptResult.errors) {
			client.logger.error(scriptResult.errors, 'contractHandler: Errors creating script')
			return maybeThrowOnFail(_params.throwOnFail ?? defaultThrowOnFail, {
				// TODO type errors better in callHandler
				errors: /** @type {any}*/ (scriptResult.errors),
				executionGasUsed: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				rawData: '0x',
			})
		}

		if (scriptResult.address !== undefined) {
			// TODO this isn't clean that we are mutating here
			evmInput.to = createAddress(scriptResult.address)
			_params.to = scriptResult.address
		}

		// start getting opstack information right away in parallel
		const l1FeeInfoPromise =
			vm.common.sourceId !== undefined
				? getL1FeeInformationOpStack(evmInput.data ?? numberToBytes(0), vm).catch((e) => {
						client.logger.warn(e, 'Unable to get l1 fee estimation')
						return {}
					})
				: Promise.resolve({})

		/**
		 * ************
		 * 2 RUN THE EVM
		 * ************
		 */
		client.logger.debug(
			{
				to: evmInput.to?.toString(),
				origin: evmInput.origin?.toString(),
				caller: evmInput.caller?.toString(),
				value: evmInput.value?.toString(),
				gasLimit: evmInput.gasLimit?.toString(),
				data: evmInput.data,
			},
			'callHandler: Executing runCall with params',
		)
		const executedCall = await executeCall({ ...client, getVm: () => Promise.resolve(vm) }, evmInput, _params)
		if ('errors' in executedCall) {
			return maybeThrowOnFail(_params.throwOnFail ?? defaultThrowOnFail, {
				executionGasUsed: /** @type {any}*/ (executeCall).rawData ?? 0n,
				/**
				 * @type {`0x${string}`}
				 */
				rawData: /** @type {any}*/ (executedCall).rawData ?? '0x',
				errors: executedCall.errors,
				...('runTxResult' in executedCall && executedCall.runTxResult !== undefined
					? callHandlerResult(executedCall.runTxResult, undefined, executedCall.trace, executedCall.accessList)
					: {}),
				...('trace' in executedCall && executedCall.trace !== undefined ? { trace: executedCall.trace } : {}),
				...('trace' in executedCall && executedCall.trace !== undefined ? { trace: executedCall.trace } : {}),
			})
		}
		executedCall.runTxResult

		/**
		 * ****************************
		 * 3 CREATE TRANSACTION WITH CALL (if neessary)
		 * ****************************
		 */
		const txResult = await handleTransactionCreation(client, params, executedCall, evmInput)
		if (txResult.errors) {
			return maybeThrowOnFail(_params.throwOnFail ?? defaultThrowOnFail, {
				...(vm.common.sourceId !== undefined ? await l1FeeInfoPromise : {}),
				...callHandlerResult(executedCall.runTxResult, txResult.hash, executedCall.trace, executedCall.accessList),
				errors: txResult.errors,
			})
		}

		/**
		 * ******************
		 * 4 RETURN CALL RESULT
		 * ******************
		 */
		return maybeThrowOnFail(_params.throwOnFail ?? defaultThrowOnFail, {
			...(vm.common.sourceId !== undefined ? await l1FeeInfoPromise : {}),
			...callHandlerResult(executedCall.runTxResult, txResult.hash, executedCall.trace, executedCall.accessList),
		})
	}
