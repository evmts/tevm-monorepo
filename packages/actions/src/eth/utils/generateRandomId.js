/**
 * @returns {import("@tevm/utils").Hex}
 */
export const generateRandomId = () => {
	return `0x${Array.from(crypto.getRandomValues(new Uint8Array(16)))
		.map((b) => b.toString(16).padStart(2, '0'))
		.join('')}`
}
