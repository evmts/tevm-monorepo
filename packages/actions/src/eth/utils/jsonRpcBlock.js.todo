import { bytesToHex } from "viem"

const ethjsTxToTevmTx = (tx) => {

}

/**
 * Transoforms an ethjs block to a tevm Block
 * @param {import("@ethereumjs/block").Block} block
 * @param {boolean} includeTransactions
 * @returns {Promise<import("@tevm/actions-types").Block>}
 */
const ethjsToTevmBlock = async (
  block,
  includeTransactions,
) => {
  const transactions = block.transactions.map((tx, txIndex) =>
    includeTransactions ? ethjsTxToTevmTx(tx, block, txIndex) : bytesToHex(tx.hash())
  )
  const withdrawalsAttr =
    block.header.withdrawalsRoot !== undefined
      ? {
        withdrawalsRoot: bytesToHex(block.header.withdrawalsRoot),
        withdrawals: block.toJSON().withdrawals,
      }
      : {}
  const td = await chain.getTd(block.hash(), block.header.number)
  return {
    coinbase: block.toJSON().header?.coinbase
    number: block.toJSON().header.number,
    hash: bytesToHex(block.hash()),
    parentHash: block.toJSON().header.parentHash,
    mixHash: block.toJSON().header.mixHash,
    nonce: block.toJSON().header.nonce,
    sha3Uncles: block.toJSON().header.uncleHash,
    logsBloom: block.toJSON().header.logsBloom,
    transactionsRoot: block.toJSON().header.transactionsTrie,
    stateRoot: block.toJSON().header.stateRoot,
    receiptsRoot: block.toJSON().header.receiptTrie,
    miner: block.toJSON().header.coinbase,
    difficulty: block.toJSON().header.difficulty,
    totalDifficulty: bigIntToHex(td),
    extraData: block.toJSON().header.extraData,
    size: intToHex(utf8ToBytes(JSON.stringify(block.toJSON())).byteLength),
    gasLimit: block.toJSON().header.gasLimit,
    gasUsed: block.toJSON().header.gasUsed,
    timestamp: block.toJSON().header.timestamp,
    transactions,
    uncles: block.uncleHeaders.map((uh) => bytesToHex(uh.hash())),
    baseFeePerGas: block.toJSON().header.baseFeePerGas,
    ...withdrawalsAttr,
    blobGasUsed: block.toJSON().header.blobGasUsed,
    excessBlobGas: block.toJSON().header.excessBlobGas,
    parentBeaconBlockRoot: block.toJSON().header.parentBeaconBlockRoot,
  }
}
