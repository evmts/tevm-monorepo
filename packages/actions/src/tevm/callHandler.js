import { callHandlerOpts } from './callHandlerOpts.js'
import { callHandlerResult } from './callHandlerResult.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { validateCallParams } from '@tevm/zod'

/**
 * Creates an CallHandler for handling call params with Ethereumjs EVM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'vm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail]
 * @returns {import('@tevm/actions-types').CallHandler}
 */
export const callHandler =
	(client, { throwOnFail: defaultThrowOnFail = true } = {}) =>
	async (params) => {
		/**
		 * @type {import('@tevm/vm').TevmVm}
		 */
		let copiedVm

		try {
			copiedVm = params.createTransaction
				? client.vm
				: await client.vm.deepCopy()
		} catch (e) {
			return maybeThrowOnFail(params.throwOnFail ?? defaultThrowOnFail, {
				errors: [
					{
						name: 'UnexpectedError',
						_tag: 'UnexpectedError',
						message:
							typeof e === 'string'
								? e
								: e instanceof Error
								? e.message
								: 'unknown error',
					},
				],
				executionGasUsed: 0n,
				rawData: '0x',
			})
		}

		const validationErrors = validateCallParams(params)
		if (validationErrors.length > 0) {
			return maybeThrowOnFail(params.throwOnFail ?? defaultThrowOnFail, {
				errors: validationErrors,
				executionGasUsed: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				rawData: '0x',
			})
		}

		try {
			const evmResult = await copiedVm.evm.runCall(callHandlerOpts(params))
			if (params.createTransaction && !evmResult.execResult.exceptionError) {
				copiedVm.stateManager.checkpoint()
				copiedVm.stateManager.commit()
			}
			return /** @type {any}*/ (
				maybeThrowOnFail(
					params.throwOnFail ?? defaultThrowOnFail,
					callHandlerResult(evmResult),
				)
			)
		} catch (e) {
			return maybeThrowOnFail(params.throwOnFail ?? defaultThrowOnFail, {
				errors: [
					{
						name: 'UnexpectedError',
						_tag: 'UnexpectedError',
						message:
							typeof e === 'string'
								? e
								: e instanceof Error
								? e.message
								: 'unknown error',
					},
				],
				executionGasUsed: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				rawData: '0x',
			})
		}
	}
