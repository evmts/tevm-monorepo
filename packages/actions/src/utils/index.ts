export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
