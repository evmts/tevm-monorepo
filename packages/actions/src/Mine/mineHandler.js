import { InternalError, MisconfiguredClientError, UnreachableCodeError } from '@tevm/errors'
import { bytesToHex, hexToBytes } from '@tevm/utils'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { emitEvents } from './emitEvents.js'
import { validateMineParams } from './validateMineParams.js'

// TODO Errors can leave us in bad states

/**
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('./MineHandlerType.js').MineHandler}
 */
export const mineHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params } = {}) => {
		client.logger.debug({ throwOnFail, ...params }, 'mineHandler called with params')
		const errors = validateMineParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}
		const { interval = 1, blockCount = 1 } = params

		/**
		 * @type {Array<import('@tevm/block').Block>}
		 */
		const newBlocks = []
		/**
		 * @type {Map<import('@tevm/utils').Hex,Array<import('@tevm/receipt-manager').TxReceipt>>}
		 */
		const newReceipts = new Map()

		client.logger.debug({ blockCount }, 'processing txs')
		const pool = await client.getTxPool()
		const originalVm = await client.getVm()

		switch (client.status) {
			case 'MINING': {
				// wait for the previous mine to finish
				await new Promise((resolve) => {
					client.on('newBlock', async () => {
						if (client.status === 'MINING') {
							return
						}
						client.status = 'MINING'
						resolve(client)
					})
				})
				break
			}
			case 'INITIALIZING': {
				await client.ready()
				client.status = 'MINING'
				break
			}
			case 'SYNCING': {
				const err = new MisconfiguredClientError('Syncing not currently implemented')
				return maybeThrowOnFail(throwOnFail, { errors: [err] })
			}
			case 'STOPPED': {
				const err = new MisconfiguredClientError('Client is stopped')
				return maybeThrowOnFail(throwOnFail, { errors: [err] })
			}
			case 'READY': {
				client.status = 'MINING'
				break
			}
			default: {
				const err = new UnreachableCodeError(client.status)
				return maybeThrowOnFail(throwOnFail, { errors: [err] })
			}
		}

		const vm = await originalVm.deepCopy()
		const receiptsManager = await client.getReceiptsManager()

		for (let count = 0; count < blockCount; count++) {
			const parentBlock = await vm.blockchain.getCanonicalHeadBlock()

			let timestamp = Math.max(Math.floor(Date.now() / 1000), Number(parentBlock.header.timestamp))
			timestamp = count === 0 ? timestamp : timestamp + interval

			const blockBuilder = await vm.buildBlock({
				parentBlock,
				headerData: {
					timestamp,
					number: parentBlock.header.number + 1n,
					// The following 2 are currently not supported
					// difficulty: undefined,
					// coinbase,
					gasLimit: parentBlock.header.gasLimit,
					baseFeePerGas: parentBlock.header.calcNextBaseFee(),
				},
				blockOpts: {
					// Proof of authority not currently supported
					// cliqueSigner,
					// proof of work not currently supported
					//calcDifficultyFromHeader,
					//ck
					freeze: false,
					setHardfork: false,
					putBlockIntoBlockchain: false,
					common: vm.common,
				},
			})
			// TODO create a Log manager
			const orderedTx = await pool.txsByPriceAndNonce({ baseFee: parentBlock.header.calcNextBaseFee() })

			let index = 0
			// TODO we need to actually handle this
			const blockFull = false
			/**
			 * @type {Array<import('@tevm/receipt-manager').TxReceipt>}
			 */
			const receipts = []
			while (index < orderedTx.length && !blockFull) {
				const nextTx = /** @type {import('@tevm/tx').TypedTransaction}*/ (orderedTx[index])
				client.logger.debug(bytesToHex(nextTx.hash()), 'new tx added')
				const txResult = await blockBuilder.addTransaction(nextTx, {
					skipHardForkValidation: true,
				})
				if (txResult.execResult.exceptionError) {
					if (txResult.execResult.exceptionError.error === 'out of gas') {
						client.logger.debug(txResult.execResult.executionGasUsed, 'out of gas')
					}
					client.logger.debug(
						txResult.execResult.exceptionError,
						`There was an exception when building block for tx ${bytesToHex(nextTx.hash())}`,
					)
				}
				receipts.push(txResult.receipt)
				index++
			}
			await vm.stateManager.checkpoint()
			const createNewStateRoot = true
			await vm.stateManager.commit(createNewStateRoot)
			const block = await blockBuilder.build()
			await Promise.all([receiptsManager.saveReceipts(block, receipts), vm.blockchain.putBlock(block)])
			pool.removeNewBlockTxs([block])

			newBlocks.push(block)
			newReceipts.set(bytesToHex(block.hash()), receipts)

			const value = vm.stateManager._baseState.stateRoots.get(bytesToHex(block.header.stateRoot))

			if (!value) {
				throw new InternalError(
					'InternalError: State root not found in mineHandler. This indicates a potential inconsistency in state management.',
				)
			}

			originalVm.stateManager.saveStateRoot(block.header.stateRoot, value)
		}
		originalVm.blockchain = vm.blockchain
		originalVm.evm.blockchain = vm.evm.blockchain
		await originalVm.stateManager.setStateRoot(hexToBytes(vm.stateManager._baseState.getCurrentStateRoot()))

		client.status = 'READY'

		emitEvents(client, newBlocks, newReceipts)

		return { blockHashes: newBlocks.map((b) => bytesToHex(b.hash())) }
	}
