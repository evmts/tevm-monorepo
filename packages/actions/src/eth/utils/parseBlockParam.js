import { InvalidBlockError } from '@tevm/errors'
import { hexToBigInt, hexToBytes } from '@tevm/utils'

/**
 * @param {import('@tevm/blockchain').Chain} blockchain
 * @param {import('../../common/BlockParam.js').BlockParam} blockParam
 * @returns {Promise<bigint >}
 */
export const parseBlockParam = async (blockchain, blockParam) => {
	if (typeof blockParam === 'number') {
		return BigInt(blockParam)
	}
	if (typeof blockParam === 'bigint') {
		return blockParam
	}
	if (typeof blockParam === 'string' && blockParam.startsWith('0x')) {
		if (blockParam.length === 66) {
			const block = await blockchain.getBlock(hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (blockParam)))
			return BigInt(block.header.number)
		}
		return hexToBigInt(/** @type {import('@tevm/utils').Hex}*/ (blockParam))
	}
	if (blockParam === 'safe') {
		const safeBlock = blockchain.blocksByTag.get('safe')
		// let's handle it here in case we forget to update it later
		if (safeBlock) {
			return safeBlock.header.number
		}
		throw new InvalidBlockError('safe not currently supported as block tag')
	}
	if (blockParam === 'latest' || blockParam === undefined) {
		const safeBlock = blockchain.blocksByTag.get('latest')
		// let's handle it here in case we forget to update it later
		if (safeBlock) {
			return safeBlock.header.number
		}
		throw new InvalidBlockError('latest block does not exist on chain')
	}
	if (blockParam === 'pending') {
		// for pending we need to mine a new block and then handle it
		// let's skip this functionality for now
		throw new InvalidBlockError(
			'Pending not yet supported but will be in future. Consider opening an issue or reaching out on telegram if you need this feature to expediate its release',
		)
	}
	if (blockParam === 'earliest') {
		return BigInt(1)
	}
	if (blockParam === 'finalized') {
		throw new InvalidBlockError('finalized noet yet supported for this feature')
	}
	blockchain.logger.error({ blockParam }, 'Unknown block param pased to blockNumberHandler')
	throw new InvalidBlockError(`Unknown block param ${blockParam} pased to blockNumberHandler`)
}
