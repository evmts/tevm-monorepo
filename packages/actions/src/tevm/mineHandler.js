import { bytesToHex } from '@tevm/utils'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { validateMineParams } from '@tevm/zod'

// TODO Errors can leave us in bad states

/**
 * @param {import("@tevm/base-client").BaseClient} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').MineHandler}
 */
export const mineHandler =
  (client, options = {}) =>
    async ({ throwOnFail = options.throwOnFail ?? true, ...params } = {}) => {
      client.logger.debug({ throwOnFail, ...params }, 'mineHandler called with params')
      const errors = validateMineParams(params)
      if (errors.length > 0) {
        return maybeThrowOnFail(throwOnFail, { errors })
      }
      const { interval = 1, blockCount = 1 } = params

      /**
       * @type {Array<import('@tevm/utils').Hex>}
       */
      const blockHashes = []

      client.logger.debug({ blockCount }, 'processing txs')

      for (let count = 0; count < blockCount; count++) {
        const pool = await client.getTxPool()
        const vm = await client.getVm()
        const receiptsManager = await client.getReceiptsManager()
        const parentBlock = await vm.blockchain.getCanonicalHeadBlock()

        let timestamp = Math.max(Math.floor(Date.now() / 1000), Number(parentBlock.header.timestamp))
        timestamp = count === 0 ? timestamp : timestamp + interval

        const blockBuilder = await vm.buildBlock({
          parentBlock,
          headerData: {
            timestamp,
            number: parentBlock.header.number + 1n,
            // The following 2 are currently not supported
            // difficulty: undefined,
            // coinbase,
            gasLimit: parentBlock.header.gasLimit,
            baseFeePerGas: parentBlock.header.calcNextBaseFee(),
          },
          blockOpts: {
            // Proof of authority not currently supported
            // cliqueSigner,
            // proof of work not currently supported
            //calcDifficultyFromHeader,
            //ck
            freeze: false,
            setHardfork: false,
            putBlockIntoBlockchain: false,
            common: vm.common
          },
        })
        // TODO create a Log manager
        const orderedTx = await pool.txsByPriceAndNonce({ baseFee: parentBlock.header.calcNextBaseFee() })

        let index = 0
        let blockFull = false
        /**
         * @type {Array<import('@tevm/receipt-manager').TxReceipt>}
         */
        const receipts = []
        while (index < orderedTx.length && !blockFull) {
          const nextTx = /** @type {import('@tevm/tx').TypedTransaction}*/(orderedTx[index])
          client.logger.debug(bytesToHex(nextTx.hash()), 'new tx added')
          const txResult = await blockBuilder.addTransaction(nextTx, {
            skipHardForkValidation: true,
          })
          if (txResult.execResult.exceptionError) {
            if (txResult.execResult.exceptionError.error === 'out of gas') {
              client.logger.debug(txResult.execResult.executionGasUsed, 'out of gas')
            }
            client.logger.debug(txResult.execResult.exceptionError, `There was an exception when building block for tx ${bytesToHex(nextTx.hash())}`)
          }
          receipts.push(txResult.receipt)
          index++
        }
        await vm.stateManager.checkpoint()
        const createNewStateRoot = true
        await vm.stateManager.commit(createNewStateRoot)
        const block = await blockBuilder.build()
        await Promise.all([
          receiptsManager.saveReceipts(block, receipts),
          vm.blockchain.putBlock(block),
        ])
        pool.removeNewBlockTxs([block])

        blockHashes.push(bytesToHex(block.hash()))
      }
      return { blockHashes }
    }
