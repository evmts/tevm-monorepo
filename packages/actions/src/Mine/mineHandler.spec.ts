import { type TevmNode, createTevmNode } from '@tevm/node'
import { type Hex, hexToBytes } from '@tevm/utils'
import { http } from 'viem'
import { describe, expect, it } from 'vitest'
import { callHandler } from '../Call/callHandler.js'
import { mineHandler } from './mineHandler.js'

const getBlockNumber = (client: TevmNode) => {
	return client
		.getVm()
		.then((vm) => vm.blockchain.getCanonicalHeadBlock())
		.then((block) => block.header.number)
}

describe(mineHandler.name, () => {
	it('as a default it should mine 1 block', async () => {
		const client = createTevmNode()
		expect(await getBlockNumber(client)).toBe(0n)
		await mineHandler(client)({})
		expect(await getBlockNumber(client)).toBe(1n)
	})

	it(
		'should work in forked mode too',
		async () => {
			const client = createTevmNode({ fork: { transport: http('https://mainnet.optimism.io')({}) } })
			const bn = await getBlockNumber(client)
			expect(bn).toBeGreaterThan(119504797n)
			await mineHandler(client)({})
			expect(await getBlockNumber(client)).toBe(bn + 1n)
		},
		{ timeout: 20_000 },
	)

	it('can be passed blockCount and interval props', async () => {
		const client = createTevmNode()
		expect(await getBlockNumber(client)).toBe(0n)
		await mineHandler(client)({
			interval: 2,
			blockCount: 2,
		})
		expect(await getBlockNumber(client)).toBe(2n)
		const latestBlock = await client.getVm().then((vm) => vm.blockchain.getCanonicalHeadBlock())
		expect(latestBlock.header.number).toBe(2n)
		const parentBlock = await client.getVm().then((vm) => vm.blockchain.getBlock(latestBlock.header.parentHash))
		expect(parentBlock.header.timestamp + 2n).toBe(latestBlock.header.timestamp)
	})

	it('works with transactions in the tx pool', async () => {
		const client = createTevmNode()
		const to = `0x${'69'.repeat(20)}` as const
		// send value
		const callResult = await callHandler(client)({
			createTransaction: true,
			to,
			value: 420n,
			skipBalance: true,
		})
		expect(callResult).toEqual({
			executionGasUsed: 0n,
			rawData: '0x',
			txHash: '0x5e5b342fae6b13548e62c3038078915397ebd2406a8c67afd276e8dc84ebba80',
			amountSpent: 147000n,
			totalGasSpent: 21000n,
		})
		expect(await client.getTxPool().then((pool) => [...pool.pool.keys()].length)).toBe(1)
		const { blockHashes, errors } = await mineHandler(client)({})

		expect(errors).toBeUndefined()
		expect(blockHashes).toHaveLength(1)

		expect(await getBlockNumber(client)).toBe(1n)

		// receipt should exist now
		const receiptsManager = await client.getReceiptsManager()
		// const block = await (await client.getVm()).blockchain.getCanonicalHeadBlock()
		const receipt = await receiptsManager.getReceiptByTxHash(hexToBytes(callResult.txHash as Hex))

		if (receipt === null) throw new Error('Receipt is null')

		expect(receipt[0].logs).toMatchSnapshot()
		expect(receipt[0].bitvector).toMatchSnapshot()
		expect(receipt[0].cumulativeBlockGasUsed).toMatchSnapshot()
		expect(receipt[1]).toBeDefined()
		expect(receipt[2]).toBeDefined()

		// should remove tx from mempool
		expect(await client.getTxPool().then((pool) => [...pool.pool.keys()].length)).toBe(0)
	})
})
