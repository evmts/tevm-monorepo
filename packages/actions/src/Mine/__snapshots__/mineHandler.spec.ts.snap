// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`mineHandler works with transactions in the tx pool 1`] = `[]`;

exports[`mineHandler works with transactions in the tx pool 2`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`mineHandler works with transactions in the tx pool 3`] = `21000n`;
